### DISCUSSION TITLE
supplied material check

### DISCUSSION THREAD

**Discussion by mghkorzani** [idaholab/moose#18170] [Q&A General] [answered]
Hi all,
I have a question regarding material properties.
I have a pointer material within a new material class (e.g. MatA class). I would like to see if a property name is declared by another material class  or not (e.g. MatB class). How can I check it? Because if the property is declared in MatB class, I can point my pointer material in MatA class to that property in MatB class. Otherwise, I should locally calculate the required property in my MatA class and make my pointer material null.
I appreciate your helps.
Cheers,
Maziar

**Comment by GiudGiud** [comment]
Hi Maziar

you can use getMaterialProperty from a material. I would create the material property in the constructor of the material, then check if it is defined in the current block in the compute properties routine. 

If defined, then use that value
If not use the new material s value. 

Guillaume

**Comment by mghkorzani** [comment]
Hi Guillaume,

 

Thanks for your answer.

The problem is here. I always use getMaterialProperty to read properties that are defined in other classes. However in this case, if it is not defined in another material class, I will receive an error because I tried to initialize my pointer. I tried hasMaterialProperty but did not work since it is looking for properties that are defined in its own class.

Maybe it is better to explain the real problem:

I try to read the strain rate from the tensor mechanics module. In the case of small strain, it is not declared but for incremental cases, it is declared as a ranktwotensor. I would like to see my material realise it by itself which means:

1. In small non-incremental strain (total strain), I calculate it in my material. -> I already did the calculation.

2. In incremental strain, I just get it from the tensor mechanics material property.  -> I know how to get it and I used it but I need to know in my material constructor that I should get this material or calculate it myself.

 

Cheers,

Maziar

**Comment by GiudGiud** [comment]
Hello

Ok thanks for the explanation I thought this was a block restriction issue. 

You could have a boolean in your material. Or equivalently specify the strain model to the material.
Then at construction you can use _matprop(_need_to_declare_it ? declareMatprop() : getMaterialProperty() )
Or something more detailed in the body of the constructor with a mooseError or mooseWarning if not enough information is provided by the user to make the selection.

If you try to retrieve a material property and it doesnt exist, it will error out. I dont think there is a way around it, but @rwcarlsen would know more about the material system. 

Guillaume

**Comment by rwcarlsen** [comment, marked as answer]
Normally `hasMaterialProperty` would do what you want - e.g. something like `MaterialProperty<Foo> * foo = hasMaterialProperty(...) ? getMaterialProperty(...) : nullptr`.  The problem is that you are doing this within a material object's constructor - and so all the materials might not be constructed yet - so you might get an error even though that property will be defined by another material after.  Also, if you call getMaterialProperty in a different function that runs later (e.g. not the constructor) - then you risk moose not seeing the inter-property dependency.  One workaround might be putting your getMaterialProperty call in the material's constructor body - and surround it with a try-catch.  In the catch - you just make the property be a nullptr and then you know you need to calculate it manually.

**Comment by mghkorzani** [comment]
Thanks for the help.I set a bool input parameter to be supplied by users with the same name of the tensor mechanics equivalent. So I can set it for the both in my global parameter. Or I should use the approach of the dictator uo as per porous flow module.Have a nice weekend.Sent from my Galaxy
-------- Original message --------From: Robert Carlsen ***@***.***> Date: 25/6/21  21:55  (GMT+01:00) To: idaholab/moose ***@***.***> Cc: Maziar Korzani ***@***.***>, Author ***@***.***> Subject: Re: [idaholab/moose] supplied material check (#18170) 
Normally hasMaterialProperty would do what you want - e.g. something like MaterialProperty<Foo> * foo = hasMaterialProperty(...) ? getMaterialProperty(...) : nullptr.  The problem is that you are doing this within a material object's constructor - and so all the materials might not be constructed yet - so you might get an error even though that property will be defined by another material after.  Also, if you call getMaterialProperty in a different function that runs later (e.g. not the constructor) - then you risk moose not seeing the inter-property dependency.  One workaround might be putting your getMaterialProperty call in the material's constructor body - and surround it with a try-catch.  In the catch - you just make the property be a nullptr and then you know you need to calculate it manually.
